modules/imgproc/include/opencv2/imgproc.hpp

```diff
@@ -4777,6 +4777,76 @@ Point LineIterator::pos() const
 
 //! @endcond
 
+class CV_EXPORTS LineVirtualIterator
+{
+public:
+    /** @brief initializes the iterator
+
+    creates iterators for the line connecting pt1 and pt2
+    the line will be clipped on the image boundaries
+    the line is 8-connected or 4-connected
+    If forceLeftToRight=true, then the iteration is always done
+    from the left-most point to the right most,
+    not to depend on the ordering of pt1 and pt2 parameters
+    */
+    LineVirtualIterator(const Point& pt1, const Point& pt2,
+                        int connectivity = 8, bool forceLeftToRight = false );
+    LineVirtualIterator(const Size& boundingAreaSize, const Point& pt1, const Point& pt2,
+                        int connectivity = 8, bool forceLeftToRight = false );
+    LineVirtualIterator(const Rect& boundingAreaRect, const Point& pt1, const Point& pt2,
+                        int connectivity = 8, bool forceLeftToRight = false );
+
+    /** @brief alias of pos(), returns coordinates of the current pixel
+    */
+    const Point& operator *() const {return pos();}
+    /** @brief prefix increment operator (++it). shifts iterator to the next pixel
+    */
+    LineVirtualIterator& operator ++();
+    /** @brief postfix increment operator (it++). shifts iterator to the next pixel
+    */
+    LineVirtualIterator operator ++(int);
+    /** @brief returns coordinates of the current pixel
+    */
+    const Point& pos() const {return currentPos;}
+
+    Size size;
+    Point currentPos;
+    Point currentPosOffset;
+    int err, count;
+    int minusDelta, plusDelta;
+    int minusStep, plusStep;
+};
+
+//! @cond IGNORED
+
+// === LineVirtualIterator implementation ===
+
+inline
+LineVirtualIterator& LineVirtualIterator::operator ++()
+{
+    if (size.width)
+    {
+      const int mask = err < 0 ? -1 : 0;
+      err += minusDelta + (plusDelta & mask);
+      const int offset = minusStep + (plusStep & mask);
+      const size_t flattenedCoord = (currentPos.y - currentPosOffset.y) * size.width + (currentPos.x - currentPosOffset.x) + offset;
+      currentPos.y = static_cast<int>((flattenedCoord / size.width) + currentPosOffset.y);
+      currentPos.x = static_cast<int>((flattenedCoord % size.width) + currentPosOffset.x);
```

Comment:
basically, I mean that this 2 lines should be removed; instead, you should update currentPos directly using the Bresenham algorithm

Reply:
The current implementation is rather far from standard Bresenham's algorithm descriptions, so it is not that straightfoward to rewrite differently all the cases in both 4 and 8 connectivity.